### ISSUE TITLE
Feature request: ElasticSearch backend compatibility mode

### ISSUE THREAD

**Issue by wkoszek** [typesense/typesense#1383]
## Description

We have developed a platform for de-identifying, storing and searching clinical narrative reports. Example can be: report from the radiology department. We have chosen early on to utilize ElasticSearch. Our frontend team used Elastic's JS widgets and modified them. We feel like frontend part is working fine, but ElasticSearch is slow. We want to get the search to be consistent and below 1-2s. We were hoping that moving to Typesense's backend would solve the speed issue, but we'd love to not throw the frontend code away.

It'd be nice if Typesense had ElasticSearch compatibility mode - the backend endpoint where queries from ES JS widgets could be sent, and handled, as if Typesense's backend was ElasticSearch. This could be a nice way for you to transparently switch existing customers to Typesense, as I imagine there are many frontends that format JSON queries according to ES.

## Steps to reproduce

We are using those JS widgets:

https://www.npmjs.com/package/@elastic/react-search-ui
https://npmjs.com/package/@elastic/search-ui-app-search-connector

## Expected Behavior

Would love to just move from e.g.: elastic.mysearchplatform.com to typesense.mysearchplatform.com transparently without rewriting anything in FE and get 10x the speed.

## Actual Behavior

We'd have to rewrite a lot of code to Typesense.

## Metadata

**Typesense Version**:

**OS**:

**Comment by wkoszek** [comment]
We (https://www.segmed.ai) would appreciate the difficulty estimate on this ticket. We were wondering if we could help sponsoring this ^^ issue and get something good enough for our purposes; we'd be fine joining forces with other potential sponsors too.

**Comment by jasonbosco** [comment]
@wkoszek I took a brief look at Search UI and it looks like they already have a way to make a call to any API to fetch search results: https://docs.elastic.co/search-ui/guides/building-a-custom-connector

In that example, they have this line:

```js
await fetch("https://api.my-host/search")
``

You'd essentially replace that with a call to Typesense's search endpoint either using `fetch` or using `typesense-js`. So it sounds to me like an "adapter" is not necessarily needed here, since the project already supports making calls to any search API endpoints...

**Comment by wkoszek** [comment]
@jasonbosco Thanks for replying so promptly. If my understanding of this extension is correct, the one line that you've listed is indeed easy to change. However I think the top reason for this issue is this line:

```
      "https://api.my-host/autocomplete?query" + state.searchTerm,
```

So this JSON blob in `state.searchTerm` will be formatted accordingly for the Elastic REST API.  Are you indicating that this JSON would be acceptable by Typesense REST API?

**Comment by jasonbosco** [comment]
> Are you indicating that this JSON would be acceptable by Typesense REST API?

No, you would have to map that object into the set of search parameters that Typesense supports: https://typesense.org/docs/0.25.1/api/search.html#search-parameters

If you can do a `console.log` of the full state object, I can help identify what the equivalent search parameters in Typesense are.

**Comment by wkoszek** [comment]
@jasonbosco OK, so you're suggesting doing the converter of JSON format from Elastic's to Typesense's in JS itself.  We could do this, but I think it'd mean work to develop and them continuous work to maintain this for our and all teams utilizing Elastic's widgets.

The idea here was that one could "transparently" switch to Typesense. I'd be OK with Typesense coming up with a JS library to do this ES JSON -> Typesense JSON converting too. But as you showed earlier - the perfect case is that you just change the URL, and it should work.

**Comment by jasonbosco** [comment]
> OK, so you're suggesting doing the converter of JSON format from Elastic's to Typesense's in JS itself

That's correct. This is what we do for eg with Algolia's Search UI library (Instantsearch). We maintain that JS-based transformation as a [separate library](https://github.com/typesense/typesense-instantsearch-adapter) though, because it's quite involved and several users have asked us for it. 

> but I think it'd mean work to develop and them continuous work to maintain this for our and all teams utilizing Elastic's widgets.

If you could give me the JSON object that's generated by Elastic's Search UI, I can see how complex or simple the transformation is...

**Comment by lucasgreenwell** [comment]
Hi Jason, I'm one of the devs at Segmed. Thanks for your prompt feedback on this request. Here's a sample of a "blank" request that we send to ES. Obviously this JSON changes based on queries and filters but it has a similar structure. 
[blankESRequest.json](https://github.com/typesense/typesense/files/13395078/blankESRequest.json)
